src/commonMain/kotlin/org/kbson/serialization/Serializers.kt

```diff
@@ -0,0 +1,799 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kbson.serialization
+
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.Serializer
+import kotlinx.serialization.builtins.ListSerializer
+import kotlinx.serialization.builtins.MapSerializer
+import kotlinx.serialization.builtins.serializer
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonArray
+import kotlinx.serialization.json.JsonDecoder
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.JsonEncoder
+import kotlinx.serialization.json.JsonNull
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.booleanOrNull
+import kotlinx.serialization.json.contentOrNull
+import kotlinx.serialization.json.doubleOrNull
+import kotlinx.serialization.json.floatOrNull
+import kotlinx.serialization.json.intOrNull
+import kotlinx.serialization.json.longOrNull
+import kotlinx.serialization.modules.SerializersModule
+import org.kbson.BsonArray
+import org.kbson.BsonBinary
+import org.kbson.BsonBoolean
+import org.kbson.BsonDBPointer
+import org.kbson.BsonDateTime
+import org.kbson.BsonDecimal128
+import org.kbson.BsonDocument
+import org.kbson.BsonDouble
+import org.kbson.BsonInt32
+import org.kbson.BsonInt64
+import org.kbson.BsonJavaScript
+import org.kbson.BsonJavaScriptWithScope
+import org.kbson.BsonMaxKey
+import org.kbson.BsonMinKey
+import org.kbson.BsonNull
+import org.kbson.BsonObjectId
+import org.kbson.BsonRegularExpression
+import org.kbson.BsonSerializationException
+import org.kbson.BsonString
+import org.kbson.BsonSymbol
+import org.kbson.BsonTimestamp
+import org.kbson.BsonType
+import org.kbson.BsonUndefined
+import org.kbson.BsonValue
+import org.kbson.internal.Base64Utils
+import org.kbson.internal.HexUtils
+import org.kbson.internal.validateSerialization
+
+/**
+ * The BsonSerializersModule
+ *
+ * Contains serializers that handle the conversion of Bson data to extended json and back
+ */
+public val BsonSerializersModule: SerializersModule = SerializersModule {
+    contextual(BsonValue::class, BsonValueSerializer)
+}
+
+
+/**
+ * The Bson companion object
+ */
+public object Bson {
```

Comment:
Are you planning to add support for generating and parsing relaxed extended JSON?  That seems to be implied by the scope goals.

Reply:
No plans to at this time. As the goals only talk about passing extended Json strings and we only discussed using Strict encoding:

> Support serialization between BsonDocument and JSON, as defined by [Extended JSON 2.2.0 specification](https://github.com/mongodb/specifications/blob/master/source/extended-json.rst).
> The Realm Kotlin SDK needs a way to send BSON across the JNI boundary to Realm Core. The most straightforward way seems to be if BSON documents can be serialized to Extended JSON strings.

If there is a future need for different types of serialization then future refactorings will be needed.

Reply:
Both Relaxed and Strict modes are defined in the extended JSON specification, but if we only need Strict that's fine for now.  I don't recall discussing it though. @cmelchior does this work for you?

Reply:
@cmelchior we still need an answer here, but let's not hold up this PR on a decision.  If we need relaxed JSON we can implement in a subsequent PR.

Reply:
Since we are only using extended JSON as a debugging mechanism and as an internal protocol. Just having strict support should be fine 👍